### DISCUSSION TITLE
Managing user data (auth.users & profiles tables)

### DISCUSSION THREAD

**Discussion by dilly-sixmo** [supabase/supabase#9238] [Questions]
I know there's already a lot of discussion on this topic as well as this guide [here](https://supabase.com/docs/guides/auth/managing-user-data), but I'm still unsure of the ideal way to store and query user data.  Let's break it down.

1. When a new user signs up, a new entry in `auth.users` is created.  This table has fields for the user's `email`, `phone`, among others.
2. A function can be triggered on user sign up to add more user data (e.g., `first_name`, `last_name`, etc) to another table (e.g., `profiles`)  This table is related to `auth.users` with a `user_id` field.
3. Additional tables (e.g., `employees`) may be created where each entry is also related to a specific `auth.user` with `user_id` field.

Now let's say I have a page on my app (`/employees`) that pulls down and displays all the records in the `employees` table.  I want to display the employee data, as well as the employees name (stored in `profiles`) and the employee's email (stored in `auth.users`).

As far as I can tell, this is not entirely possible, because you can't query the related `auth.users` table.  (e.g., something like `.select('*, user:user_id ( * )')` isn't possible)

So this leads me to the question.  How do I display an employee's email if the `auth.users` table is storing it and it can't be queried when selecting `employees`.  Must I duplicate this `email` field in either the `profiles` or `employees` table, creating data redundancy and now forcing me to have to manage the user's `email` in two different places?

**Comment by dilly-sixmo** [comment]
I'm going to share my solution to this problem, but I'm not sure if it's the ideal solution.  If it is let me know, but if this can be improved somehow I'd love to know how.

1. `profiles` contains its own `email` field
2. When a new `profile` is created (triggered by user addition), `profile.email` is set to `user.email` (duplicating data)
3. The primary key for `profile` is `user_id`
4. Instead of relating each `employee` to the `auth.users` table using `user_id`, I can relate each `employee` to a `profile` using `profile_id` (which == `user_id`).
5. Now, when I query `employees`, I can do `.select('*, profile:profile_id (*)'` and I can get each employee's data & email in a single query.

Since you can't pull any data from `auth.users` using a foreign key relationship, it would seem that relating any table to `auth.users` is pointless except for RLS purposes.

The only issue I have here is that a user's email address is stored in two places.  But from my current understanding of supabase, this is just how I have to do it.  Is this an acceptable answer?

**Reply by julianpoma** [reply]
It is the way to do it. You can keep both fields in sync with triggers.

Also take into consideration that the emails would be viewable to anyone who can query the profiles table.

**Reply by flxmrk** [reply]
Yea, its that consideration that has brought me here... I mean, do we really have to make profiles SELECT wide open to authenticated users just because we'll want to be able to pull their name to show to a different user?  This seems nuts! Because then any authenticated user can just pull your whole user database .. its so nuts i almost think I'm just not understanding something... which is likely as I'm quite new to BaaS thinking...

**Reply by dilly-sixmo** [reply]
I made my `profiles` table private.  I'm just using it as an accessible extension of the `auth.users` table and not as a public facing profile.

**Reply by flxmrk** [reply]
@dilly-sixmo yea, that works if you never need to show the profile info to anyone but the profile owner... but I think its more often the case that profile info DOES need shown to others.  Something you'd normally just control easy in your backend programming... 

and further, i wonder, are there policies that allow only some columns of data to be read? or is it the whole kaboodle or nothing?

**Reply by julianpoma** [reply]
I think is all or nothing. The profiles table is just an extension of `auth.users`, as said above. You should only store the bare minimum data. I have many apps where emails are not in the `profiles` table. Most of the time is just an username and some user configuration stuff.
